### ISSUE TITLE
C++ optional applied to interfaces sourced from yaml

### ISSUE THREAD

**Issue by msjarrett** [dropbox/djinni#401]
### myint.djinni
```
myint = interface +c {...}
```

### myrec.djinni
```
myrec = record {
   inty: optional<myint>;
}
```

If myrec.djinni uses `@import "myint.djinni"`, it generates `std::shared_ptr<::Myint>> inty`, but if it uses `@extern "myint.yaml"` it generates `std::optional<std::shared_ptr<::Myint>> inty`

**Comment by msjarrett** [comment]
Realized that record { field: interface} is actually invalid, which the optional<interface> hides. But you get the exact same generation discrepancy with a method parameter as well.

```
myint2 = interface +c {
    dostuff(foo: optional<inty>);
}
```

**Comment by artwyman** [comment]
I agree this sounds like a bug.  We shouldn't ever generate optional<shared_ptr<T>> but just shared_ptr<T> or nn<shared_ptr<T>>.

**Comment by ctin** [comment]
https://github.com/dropbox/djinni/pull/420

**Comment by ctin** [comment]
@msjarrett fixed in https://github.com/dropbox/djinni/pull/420, so you can for and apply if need it not

**Comment by artwyman** [comment]
Fixed by #420.
